### PR TITLE
Feature/unr 4028 has authority tests

### PR DESCRIPTION
**Pull request by nafonso**
#### Description
Test for HasAuthority checks, supports Spatial and Native.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+  * in BeginPlay and Tick, and in the following use cases:
+  *		- replicated level actor
+  *		- non-replicated level actor
+  *		- dynamic replicated actor (one time with spatial authority and another without)
+  *		- dynamic non-replicated actor
+  *		- GameMode (which only exists on Servers)
+  *		- GameState
+  * Keep in mind that we're assuming a 1x2 Grid Load-Balancing Strategy, otherwise the ownership of
+  * these actors may be something completely different (specially important for actors placed in the Level).
+  * You have some flexibility to change the Server1/2Position properties to test in different Load-Balancing Strategies.
```
Comment (AndreiTeodorLazar) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp:23]: Super NIT
```suggestion
  * You have some flexibility to change the Server1/2 Position properties to test in different Load-Balancing Strategies.
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+					if (DynamicNonReplicatedActor->AuthorityOnBeginPlay == 1
+						&& DynamicNonReplicatedActor->AuthorityOnTick == 1)
+					{
+						FinishStep();
+					}
+				});
+
+		float Timer = 0.5f;
+
+		AddStep(TEXT("Non-replicated Dynamic Actor - Verify Dynamic Actor doesn't exist on others"), FWorkerDefinition::AllWorkers, nullptr, nullptr,
```
Comment (AndreiTeodorLazar) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp:162]: In this step, why do you use a Timer variable instead of relying on the test timeout to fail the step? Is it because you want to add the specific message if the test fails in the last lines of the step?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTestGameState.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+
+#include "SpatialAuthorityTestGameState.h"
+#include "Components/SceneComponent.h"
```
Comment (AndreiTeodorLazar) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTestGameState.cpp:5]: I might be wrong, but is this include needed? 
```suggestion

```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp (hunk 1 above)
Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp:23]: That's together because it's the name of the variable

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp (hunk 2 above)
Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp:162]: Because that timer is an active wait, we wait that time and only then we do a check

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp (hunk 1 above)
Comment (AndreiTeodorLazar) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp:23]: Oh, that makes sense, thank you!

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp (hunk 2 above)
Comment (AndreiTeodorLazar) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp:162]: So basically you try to avoid situations like when the test could pass before the timer elapsed, even though it should not pass, right? I think that makes sense, thank you for the clarification 😄!

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTestGameState.cpp (hunk 3 above)
Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTestGameState.cpp:5]: I think you're right, will double check

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp (hunk 2 above)
Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp:162]: Yes, this is the "worst kind of condition", because you need to actively wait to make sure that something never got propagated to you. We should avoid them in general, but some times we can't :-/ Not sure how we can better handle this, maybe at a certain point active wait tests can be separate, but it may require duplicating chunks of tests, so not great :-/

Comment (AndreiTeodorLazar) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp:162]: I see, thank you for the info!

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+  *		- GameMode (which only exists on Servers)
+  *		- GameState
+  * Keep in mind that we're assuming a 1x2 Grid Load-Balancing Strategy, otherwise the ownership of
+  * these actors may be something completely different (specially important for actors placed in the Level).
+  * You have some flexibility to change the Server1/2Position properties to test in different Load-Balancing Strategies.
+  */
+ASpatialAuthorityTest::ASpatialAuthorityTest()
+{
+	Author = "Nuno Afonso";
+	Description = TEXT("Test HasAuthority under multi-worker setups. It also ensure to work in Native");
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp:28]:
```suggestion
	Description = TEXT("Test HasAuthority under multi-worker setups. It also ensures it works in Native");
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+						&& DynamicNonReplicatedActor->AuthorityOnTick == 1)
+					{
+						FinishStep();
+					}
+				});
+
+		float Timer = 0.5f;
+
+		AddStep(TEXT("Non-replicated Dynamic Actor - Verify Dynamic Actor doesn't exist on others"), FWorkerDefinition::AllWorkers, nullptr, nullptr,
+				[this, &Timer](float DeltaTime) {
```
Comment (AndreiTeodorLazar) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp:163]: I might be very wrong about this, but Isn't this potentially causing undefined behaviour, because you capture the reference to a local variable?

Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp:163]: I was looking at this online and I think you may be right. Not a fan of this :/ I'll change it in a bit to have a class variable.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTestActor.cpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+	DOREPLIFETIME(ASpatialAuthorityTestActor, AuthorityOnBeginPlay);
+	DOREPLIFETIME(ASpatialAuthorityTestActor, AuthorityOnTick);
+}
+
+void ASpatialAuthorityTestActor::BeginPlay()
+{
+	Super::BeginPlay();
+
+	if (HasAuthority() && AuthorityOnBeginPlay == 0)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTestActor.cpp:30]: If HasAuthority returns true for multiple workers, then this test could still pass as the runtime would eventually replicate the correct state out to all interested workers.
Chatting with @mironec I think this would be better if AuthorityOnBeginPlay and AuthorityOnTick weren't replicated, but just local state. Then in each step we could check that one and only one worker has ever had this property set to true. This then also validates that HasAuthority is returning false on all other workers.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+}
+
+void ASpatialAuthorityTest::StartTest()
+{
+	GameModeServerAuthorities.SetNum(GetNumberOfServerWorkers());
+	GameStateServerAuthorities.SetNum(GetNumberOfServerWorkers());
+	GameStateClientAuthorities.SetNum(GetNumberOfClientWorkers());
+
+	// Make sure they're zero'ed for reruns to be more accurate
+	for(int i = 0; i < GameModeServerAuthorities.Num(); ++i)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp:324]: nit: use range based iterator

File: ci/setup-build-test-gdk.ps1 [hunk 8]
```diff
@@ -69,9 +69,10 @@ $tests = @()
 
 if ((Test-Path env:TEST_CONFIG) -And ($env:TEST_CONFIG -eq "Native")) {
     # We run spatial tests against Vanilla UE4
-    $tests += [TestSuite]::new($gdk_test_project, "NetworkingMap", "VanillaTestResults", "/Game/Maps/FunctionalTests/SpatialNetworkingMap", "$user_gdk_settings", $False, "$user_cmd_line_args")
+    $tests += [TestSuite]::new($gdk_test_project, "NetworkingMap", "VanillaTestResults", "/Game/Maps/FunctionalTests/CI_Fast/", "$user_gdk_settings", $False, "$user_cmd_line_args")
```
Comment (m-samiec) [inline ci/setup-build-test-gdk.ps1:72]: Not familiar with this code - but why these changes?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp [hunk 9]
```diff
@@ -0,0 +1,359 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp:1]: We're trying to move away from bug ids as folder names, please give this a descriptive name instead.

File: ci/setup-build-test-gdk.ps1 (hunk 8 above)
Comment (nafonso) [reply ci/setup-build-test-gdk.ps1:72]: This was Miron as they will have a CI_Fast and CI_Slow folders with test maps that run all commits or just nightly

**Comment by nafonso** [comment]
Asking for a new review given that it was changed a lot.

**Comment by mironec** [comment]
Probably needs to be merged with master, as it shows the log changes as well.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+				}
+			}, 5.0f);
+	}
+
+	// Replicated Dynamic Actor Spawned On Same Server. Server 1 should have Authority.
+	{
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Same Server - Spawn"), FWorkerDefinition::Server(1), nullptr,
+			[this]()
+			{
+				ASpatialAuthorityTestReplicatedActor* Actor = GetWorld()->SpawnActor<ASpatialAuthorityTestReplicatedActor>(
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:97]: How does server 1 set DynamicReplicatedActor ?

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:97]: Oh I see it's replicated, and the cross server call just ensures it's set first on the correct server

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:97]: Do we statically define which server should be authoritative over ASpatialAuthorityTest ?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+						}
+					}
+				}
+			}, 5.0f);
+
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Same Server - Destroy"), FWorkerDefinition::Server(1), nullptr,
+			[this]()
+			{
+				DynamicReplicatedActor->Destroy();
+				CrossServerSetDynamicReplicatedActor(nullptr);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:131]: Should be redundant, but I guess there's no harm

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+					if(LocalWorkerDefinition.Type == ESpatialFunctionalTestWorkerType::Server)
+					{
+						// Allow it to continue working in Native / Single worker setups.
+						if (GetNumberOfServerWorkers() > 1)
+						{
+							if (LocalWorkerDefinition.Id == 1)
+							{
+								// Note: Depending on timing / data migration, Tick may or may not have ran
+								// in Server 1, so we need to check both situations.
+								if (VerifyTestActor(DynamicReplicatedActor, 1, 1, 1, 1)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:163]: What's the tick delta for this actor? If it's not 60fps we should increase it and try to remove this edge case.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.h [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+public:
+	ASpatialAuthorityTest();
+
+	virtual void BeginPlay() override;
+
+	virtual void GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const override;
+
+	virtual void FinishStep() override { ResetTimer(); Super::FinishStep();	};
+
+	void ResetTimer() {	Timer = 0.5; };
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.h:27]:
```suggestion
	void ResetTimer() { Timer = 0.5; };
```
?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTestActor.cpp [hunk 14]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialAuthorityTestActor.h"
+//#include "Components/SceneComponent.h"
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTestActor.cpp:4]: Remove comment

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTestActor.h [hunk 15]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "CoreMinimal.h"
+#include "GameFramework/Actor.h"
+#include "SpatialAuthorityTestActor.generated.h"
+
+class USpatialAuthorityTestActorComponent;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTestActor.h:9]: This is the only forward declaration we need, can remove the others

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+				{
+					FinishTest(EFunctionalTestResult::Failed, TEXT("This test requires ASpatialAuthorityTestGameMode"));
+					return;
+				}
+
+				bool bIsStateValid;
+
+				USpatialAuthorityTestActorComponent* AuthorityComponent = GameMode->AuthorityComponent;
+				// Either it's bigger than 0 and all match (in the Authoritive Server), or all equal to zero (except for replicated).
+				if (AuthorityComponent->AuthorityOnBeginPlay > 0)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:295]: Is there a way to validate that this path is followed by exactly one worker?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTestActorComponent.h [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+	virtual void OnAuthorityGained() override;
+
+	virtual void OnAuthorityLost() override;
+
+	virtual void TickComponent(float DeltaTime, enum ELevelTick TickType, FActorComponentTickFunction *ThisTickFunction);
+	
+	UPROPERTY(Replicated)
+	int ReplicatedAuthorityOnBeginPlay = 0;
+
+	int AuthorityOnBeginPlay = 0;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTestActorComponent.h:31]: nit: can we slightly rename this variables to `AuthWorkerIdOnBeginPlay` `AuthWorkerIdOnTick` `ReplicatedAuthWorkerIdOnBeginPlay`

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp [hunk 18]
```diff
@@ -0,0 +1,387 @@
...
+					FinishTest(EFunctionalTestResult::Failed, TEXT("This test requires ASpatialAuthorityTestGameState"));
+					return;
+				}
+
+				bool bIsStateValid;
+
+				USpatialAuthorityTestActorComponent* AuthorityComponent = GameState->AuthorityComponent;
+				// Either it's bigger than 0 and all match the GameMode Authority (in the Authoritive Server),
+				// or all equal to zero (except for replicated).
+				if (AuthorityComponent->AuthorityOnBeginPlay > 0)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:333]: Likewise here, would be good to validate that exactly one worker followed this path

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp [hunk 19]
```diff
@@ -0,0 +1,387 @@
...
+				{
+					FinishStep();
+				}
+			}, 5.0f);
+	}
+
+	// GameState.
+	{
+		AddStep(
+			TEXT("GameState - Determine Authority by every Server"), FWorkerDefinition::AllWorkers, nullptr, nullptr,
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:318]:
```suggestion
			TEXT("GameState - Determine Authority by all worker"), FWorkerDefinition::AllWorkers, nullptr, nullptr,
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp [hunk 20]
```diff
@@ -0,0 +1,387 @@
...
+					bIsStateValid = AuthorityComponent->AuthorityOnBeginPlay == GameModeAuthority
+						   && AuthorityComponent->ReplicatedAuthorityOnBeginPlay == GameModeAuthority
+						   && AuthorityComponent->AuthorityOnTick == GameModeAuthority
+						   && AuthorityComponent->NumAuthorityGains == 1
+						   && AuthorityComponent->NumAuthorityLosses == 0;
+				}
+				else
+				{
+
+					bIsStateValid = AuthorityComponent->ReplicatedAuthorityOnBeginPlay != 0
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:346]:
```suggestion
					bIsStateValid = AuthorityComponent->ReplicatedAuthorityOnBeginPlay == GameModeAuthority
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp [hunk 21]
```diff
@@ -0,0 +1,387 @@
...
+
+	DOREPLIFETIME(ASpatialAuthorityTest, DynamicReplicatedActor);
+}
+
+void ASpatialAuthorityTest::CrossServerSetDynamicReplicatedActor_Implementation(ASpatialAuthorityTestReplicatedActor* Actor)
+{
+	DynamicReplicatedActor = Actor;
+}
+
+bool ASpatialAuthorityTest::VerifyTestActor(ASpatialAuthorityTestActor* Actor, int AuthorityOnBeginPlay, int AuthorityOnTick
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:375]: This could just take an AActor, and dynamically get the AuthorityComponent. That way we could use it for the GameState and GameMode checks also.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp (hunk 12 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:163]: I think this could actually be pointing to a bug - isn't the buffer period supposed to be 1s? I think we should always tick at least once on the original server before migrating - if it does not happen, I suspect that something else may be going wrong?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp (hunk 18 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:333]: Maybe checking the local worker id == GameModeAuthority?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.h [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+	GENERATED_BODY()
+
+public:
+	ASpatialAuthorityTest();
+
+	virtual void BeginPlay() override;
+
+	virtual void GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const override;
+
+	virtual void FinishStep() override { ResetTimer(); Super::FinishStep();	};
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.h:25]: Does this also get called on workers which do not have the step assigned to them? E.g. Will Step(Server(1)) also reset the timer for Server2?

**Comment by mironec** [comment]
Does this have an associated TestGyms PR? And if so, can you make the branch names for the two the same, so that CI will pick up that they should be tested together?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp (hunk 10 above)
Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:97]: No, Tests will be owned by the worker where they are placed currently.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp (hunk 11 above)
Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:131]: I like to force things to null, it's a habit specially of using these engines were things are no longer valid but are not null.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp (hunk 12 above)
Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:163]: I asked about this in one of the chats and was told that it may or may not happen. Currently in the test I was always seeing at least 1 tick (can't confirm if more), the tick rate is set to 0.0f to ensure ticks every frame (mostly put it like this because like in the case of NWX they override it for actors by default only tick every 1s). The thing that I don't know is how this behaves in multiple situations, like if the machine is running tests without proper resources, low fps, etc.

If it's not meant to happen, then I can change, currently this is the behaviour I'm seeing.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.h (hunk 13 above)
Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.h:27]: I don't know man, since clang I think things started to be very odd with spacing. Here for some reason it aligned Timer with a tab, but I can only see it when I show whitespaces:
image

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.h (hunk 22 above)
Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.h:25]: No, this is a local timer. On BeginPlay() they reset it, then every time they finish a step they also reset it (so it's ready for the next step). This was for avoiding having to add StartEvent on them just to set the timer, so the idea is the previous step always leaves the local timer in a good state.

**Comment by nafonso** [comment]
> Does this have an associated TestGyms PR? And if so, can you make the branch names for the two the same, so that CI will pick up that they should be tested together?

Yes, it's this one: https://github.com/spatialos/UnrealGDKTestGyms/pull/174

Do I need to do anything specific for the CI to pick it up and test them together? Does this work across engine , gdk, and gym branches?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp (hunk 21 above)
Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:375]: The reason it wasn't done is because all the other ones check X == Y, but when doing the GameMode / GameState they also have a check of ReplicatedAuthWorkerIdOnBeginPlay != 0. I can merge them, and just do some detection if it's gamemode / gamestate to check differently for them, but since I was doing that case separately I just put them in their own step.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp (hunk 18 above)
Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:333]: In theory, the checks there would already be valid. However, what I'll do is each one that goes through there will send a message to the test, and in the end we check that only 1 sent it for GameMode and for GameState. This is the only way I can think of that would catch the authority setup being broken and all servers would get authority.

**Comment by mironec** [comment]
> Do I need to do anything specific for the CI to pick it up and test them together? Does this work across engine , gdk, and gym branches?

I checked the CI runs, and they are correctly picking up the combo (gdk + testgyms). The only thing required is for the branches to have the same name. Unfortunately, this is a slightly different process than the engine CI, where you have to specify the branch exactly in the `unreal-engine.version` file. But this branch matching now works in the gdk, testgyms, and exampleproject, I think.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp (hunk 12 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:171]: Could you link the slack thread?

Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:171]: I'll try to find it, was already a week or 2 ago.

Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:171]: https://improbable.slack.com/archives/GHKNDKH3K/p1597319216329600

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp:171]: Discussed offline, this shouldn't be a race with the current setup, and will be updated accordingly.

**Review by m-samiec — APPROVED** [review]
Great work!

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp (hunk 12 above)
Comment (nafonso) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp]: Changed it to verify that it will always tick before migrating

**Review by mironec — APPROVED** [review]
Approving as long as CI passes, it executes all the required maps, and TestGyms PR is merged at the same time.
